frontend/src/app/titles.ts

```diff
@@ -0,0 +1,121 @@
+import { inject } from "@angular/core";
+import { ActivatedRouteSnapshot, Params, ResolveFn } from "@angular/router";
+import { Apollo, gql, TypedDocumentNode } from "apollo-angular";
+import { map } from "rxjs";
+import _ from "underscore";
+
+export const SITE_NAME = 'Lettercraft & Epistolary Performance in Medieval Europe';
+export const pageTitle = (name: string) => `${name} - ${SITE_NAME}`;
+
+/**
+ * create a resolver function that relies on a graphQL query
+ *
+ * @param queryFromParams returns a graphQL query based on the route parameters
+ * @param titleFromData transforms the result of the query into a string
+ * @returns a ResolverFn to resolve the title based on the query results
+ */
+const queryTitleResolver = <QueryData>(
+    queryFromParams: (params: Params) => TypedDocumentNode<QueryData, unknown>,
+    titleFromData: (data: QueryData) => string,
+): ResolveFn<string> => {
+    return (route: ActivatedRouteSnapshot) => {
+        const query = queryFromParams(route.params);
+        return inject(Apollo).watchQuery({ query }).valueChanges.pipe(
+            map(result => result.data),
+            map(titleFromData),
+            map(pageTitle),
+        );
+    }
+}
+
+type SourceTitleQueryData = { source?: { name: string } }
+
+const sourceTitleQuery = (params: Params) => gql<SourceTitleQueryData, unknown>(`
+    query SourceTitleQuery {
```

Comment:
Instead of your own defined types (`SourceTitleQuery` et al.) I would import the generated one (i.c. `SourceTitleQueryQuery`). You can also import the generated type for the variables if you want to replace the `unknown`.